Kernel/KBOComparator.hpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+/**
+ * @file KBOComparator.hpp
+ * Defines comparator class for KBO.
+ */
+
+#ifndef __KBOComparator__
+#define __KBOComparator__
+
+#include "Forwards.hpp"
+
+#include "Lib/Stack.hpp"
+
+#include "KBO.hpp"
+
+namespace Kernel {
+
+using namespace Lib;
+
+/**
+ * Runtime specialized KBO ordering check, based on the linear KBO check
+ * also implemented in @b KBO.
+ */
+class KBOComparator
+: public OrderingComparator
+{
+public:
+  KBOComparator(const KBO& kbo) : _kbo(kbo) {}
+
+  /** This function performs the runtime specialization and creates a comparator object. */
+  static KBOComparator* create(TermList tl1, TermList tl2, const KBO& kbo);
+
+  /** Executes the runtime specialized instructions with concrete substitution. */
+  bool check(const SubstApplicator* applicator) const;
+  vstring toString() const override;
+
+private:
+  // TODO this could be done with KBO::State
+  static void countSymbols(const KBO& kbo, DHMap<unsigned,int>& vars, int& w, TermList t, int coeff);
+
+  enum InstructionTag {
+    DATA = 0u,
+    WEIGHT = 1u,
+    COMPARE_VV = 2u,
+    COMPARE_VT = 3u,
+    COMPARE_TV = 4u,
+    SUCCESS = 5u,
+  };
+
+  /**
+   * Stores atomic instructions for KBO.
+   * We have 64 bits per instruction, and the following combinations (where ?> denotes checking >):
+   * - success
+   * - weight check:          w + |var_i o \theta|_v * nvar_i ?> 0
+   * - compare var with var:  (var1 o \theta) ?> (var2 o \theta)
+   * - compare var with term: (var o \theta) ?> (term o \theta)
+   * - compare term with var: (term o \theta) ?> (var o \theta)
+   *
+   * and the following layout of instructions:
+   * |      first instruction         |      second instruction    | ...
+   * | 3 bits     | 29 bits | 32 bits | 3 bits | 29 bits | 32 bits |
+   * | SUCCESS    |
+   * | WEIGHT     |  arity  |    w    | DATA   | var_i   | nvar_i  | ... (this block `arity` times)
+   * | COMPARE_VV |  var1   |  var2   |
+   * | COMPARE_VT |  var    | <empty> |           term             |
+   * | COMPARE_TV |  var    | <empty> |           term             |
+   */
+  struct Instruction {
+    static Instruction term(Term* t) {
+      Instruction ins;
+      ins._setTag(InstructionTag::DATA);
+      ins._setTerm(t);
+      return ins;
+    }
+    static Instruction uintUint(InstructionTag t, unsigned v1 = 0, unsigned v2 = 0) {
+      Instruction ins;
+      ins._setTag(t);
+      ins._setFirstUint(v1);
+      ins._setSecondUint(v2);
+      return ins;
+    }
+    static Instruction uintInt(InstructionTag t, unsigned v1, int v2) {
+      Instruction ins;
+      ins._setTag(t);
+      ins._setFirstUint(v1);
+      ins._setCoeff(v2);
+      return ins;
+    }
```

Comment:
What do these two functions do? Why are they different?

Reply:
The second one is used in the specific case when the second argument can be also negative. It could be removed if there is some way to cast a negative integer "bitwise" to an unsigned. Is there such a thing?

Reply:
...probably, but that sounds like one of those "spot the undefined behaviour" exercises waiting to happen! OK - a comment would be good if you're making another commit, but in any case fine.

Reply:
Looking at the recently added `setBits` and `getBits`, should we be worried about breaking on some compilers? According to [this](https://stackoverflow.com/questions/9656219/how-can-uint64-t-store-a-negative-number-isnt-it-meant-to-be-unsigned), if `uint64_t` does not fit the assigned value, the compiler copies bitwise, but this is not guaranteed. Anyways, if you feel confident that this does not break, we can remove the constructor as well.

Reply:
I think the implementation of those functions is fine (although maybe we should statically check that the type is reasonable), but casting `int` to `uint64_t` and back is probably dodgy. What behaviour do you intend with this? Will the coefficient ever be negative?

This could probably be avoided altogether by using a 32-bit bitfield and then a 32-bit regular union, but then we'd have to parameterise `BitUtils`...which is probably fine, I just didn't do it yet.

Reply:
Yes, in fact we only have a weight check instruction if there is at least one negative value somewhere, otherwise the balance would be trivially OK.

How about a `reinterpret_cast` like [here](https://stackoverflow.com/questions/30048135/efficient-way-to-bit-copy-a-signed-integer-to-an-unsigned-integer)?

Reply:
Sounds sensible - I think at least this communicates the intent well, and there are definitely other sign-conversion maybe-bugs in Vampire. I think this is the last blocker, so please push and I'll merge. :-)